MAINT: Tidy PyArray_DescrConverter

eric-wieser opened the pull request:
First part of some major cleanup to make `PyArray_DescrConverter` manageable.

Likely easiest to review commit-by-commit.

---

Follow-up will be to replace the clumsy `NPY_FAIL` / `*at = NULL` stuff internally to just use a regular return value, and write `PyArray_DescrConverter` as a thin wrapper.

eric-wieser committed:
MAINT: Extract a helper function from PyArray_DescrConverter

The `goto finish` part of this if was useless, as none of the steps in `finish` were ever executed.

eric-wieser commented:
Will add this to my pile of "things which didn't work today"... Splitting the first two commits to separate PRs, will revisit the third one once they're in.

eric-wieser committed:
MAINT: Remove unnecessary gotos

This makes the control flow easier to reason about - notably, the only case that now falls through the `if` / `else` are `bytes` objects.

eric-wieser committed:
MAINT: Invert and elide conditions to reduce nesting

`PyBytes_Check(obj)` was always true in these if statements

eric-wieser committed:
MAINT: Extract a helper to create a dtype from a string

The string parsing is split all over the place in PyArray_DescrConverter.
This commit brings all of it into one place, which makes `PyArray_DescrConverter` far easier to follow.

seberg commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -1651,71 +1565,186 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
+{
+    int check_num = NPY_NOTYPE + 10;
+    int elsize = 0;
+    char endian = '=';
```
Dunno if its actually not worse, but `elsize` does not seem to need initialization (I think) and `endian` could also be a `default:` clause (editor always complains about missing default, but not sure I agree with it.

eric-wieser replied:
This is cleaned up in a future commit.

eric-wieser replied:
Wait, no it isn't - pushed the change as an update. I've left them with initialization, but put the initialization right next to the use.

seberg commented on numpy/core/src/multiarray/descriptor.c:
```diff
@@ -1651,71 +1565,186 @@ PyArray_DescrConverter(PyObject *obj, PyArray_Descr **at)
-            }
-            else {
-                item = PyDict_GetItem(typeDict, obj);
+        if (PyBytes_Check(obj)) {
```
This check can be removed as well I think? Won't gh-15261 create a merge conflict here? (although I suppose its not a huge one).

eric-wieser replied:
gh-15261 is on hold until I fix all this

eric-wieser replied:
Good catch, check removed as part of the second commit in this patch. Also some other useless code elided.

seberg reviewed (approved):
LGTM, nice moving around much easier to read broken apart. The last if would be nice to remove, since you removed one further down? The other things I do not care about.

eric-wieser committed:
MAINT: Eliminate a repeated conversion

We already converted the pyobject to a string on a previous line, there's no need to do it again.

seberg reviewed (approved):
Great, thanks. I think one of the declaration fixups came in the wrong commit, but I don't mind. I will merge this as soon as the tests all pass I think, since it will create merge conflicts with the Py3K cleanup, but it has to happen one way or the other...

eric-wieser commented:
> I think one of the declaration fixups came in the wrong commit, 

Which one?

seberg commented:
https://github.com/numpy/numpy/pull/15265/commits/6d2faa6f66e020379e3b80bd5678db4354aa76e9 commit, the removal of the duplicate `type` and `len` is in the next commit, but honestly I do not care, its a pretty history in either case, and only a compiler warning.

eric-wieser commented:
> and only a compiler warning.

Nope, that's actually a bug in that intermediate commit. Happy to fix if you're happy to wait for CI again

seberg commented:
OK, I don't mind waiting, but really a nitpick IMO.

eric-wieser committed:
MAINT: Move declarations next to their use

eric-wieser commented:
Proof the commits are the same: https://github.com/numpy/numpy/compare/5da2c01a7ae3b9fc3701801ab16d234e5425436c..ac44095f8f8aca2574091bb32d6328d0b6bbf29a

eric-wieser commented:
CI passes, and this is an identical revision to the approved one, merging...
